numpy/random/mtrand/mvhypergeom.c

```diff
@@ -0,0 +1,246 @@
+        it is assumed that the caller has already validated the value.
+    size_t num_sample
+        The number of random variates to generate.  (This is not the size of
+        the array `sample`, because each variate requires `num_colors`
+        values.)
+    long *sample
+        The array that will hold the result.  The length of this array must
+        be the product of `num_colors` and `num_sample`.
+        The array is not initialized; it is expected that array has been
+        initialized with zeros when the function is called.
+
+    Notes
+    -----
+    Here's an example that demonstrates the idea of this algorithm.
+
+    Suppose the urn contains red, green, blue and yellow marbles.
+    Let nred be the number of red marbles, and define the quantities for
+    the other colors similarly.  The total number of marbles is
+
+        total = nred + ngreen + nblue + nyellow.
+
+    To generate a sample using rk_hypergeometric:
+
+       red_sample = rk_hypergeometric(ngood=nred, nbad=total - nred,
+                                      nsample=nsample)
+
+    This gives us the number of red marbles in the sample.  The number of
+    marbles in the sample that are *not* red is nsample - red_sample.
+    To figure out the distribution of those marbles, we again use
+    rk_hypergeometric:
+
+        green_sample = rk_hypergeometric(ngood=ngreen,
+                                         nbad=total - nred - ngreen,
+                                         nsample=nsample - red_sample)
+
+    Similarly,
+
+        blue_sample = rk_hypergeometric(
+                          ngood=nblue,
+                          nbad=total - nred - ngreen - nblue,
+                          nsample=nsample - red_sample - green_sample)
+
+    Finally,
+
+        yellow_sample = total - (red_sample + green_sample + blue_sample).
+
+    The above sequence of steps is implemented as a loop for an arbitrary
+    number of colors in the innermost loop in the code below.  `remaining`
+    is the value passed to `nbad`; it is `total - colors[0]` in the first
+    call to rk_hypergeometric(), and then decreases by `colors[j]` in
+    each iteration.  `num_to_sample` is the `nsample` argument.  It
+    starts at this function's `nsample` input, and is decreased by the
+    result of the call to rk_hypergeometric() in each iteration.
+*/
+
+int multivar_hypergeom_iterative(rk_state *state,
+                                 size_t num_colors, long *colors,
+                                 long nsample,
+                                 size_t num_sample, long *sample)
+{
+    long total;
+    size_t i;
+    int more_than_half;
+
+    /* total = sum(colors) */
+    total = 0;
+    for (i = 0; i < num_colors; ++i) {
+        total += colors[i];
+    }
+
+    more_than_half = nsample > (total / 2);
+    if (more_than_half) {
+        nsample = total - nsample;
+    }
+
+    i = 0;
+    while (i < num_sample) {
```

Comment:
Why not a `for` loop?

Reply:
Really, this is a classic `for` loop
```
for(i = 0; i < num_sample; i += num_colors) {
```

Reply:
Already fixed it in the update that I'm working on. :)  The only thing I have to do before sending the update is to address your comment about `size_t`.  I'll try to get this done later tonight (i.e. within the next 8 hours or so, for those not in my time zone).